JavaScript/Angular.md

```diff
@@ -0,0 +1,687 @@
+
+Define interface for the environment: `/src/app/interfaces/environment.interface`:
+```typescript
+export interface IEnvironment {
+  production: boolean;
+}
+```
+
+Set base values `/src/environments/environment.base.ts`:
+```typescript
+import { IEnvironment } from 'app/interfaces/environment.interface';
+
+export const baseEnvironment: IEnvironment = {
+  production: false,
+}
+```
+
+This is the default `dev` environment: `/src/environments/environment.ts`:
+```typescript
+import { baseEnvironment } from './environment.base';
+
+export const environment: IEnvironment = {
+  ...baseEnvironment,
+  // Nothing is overriden, all will be as defined in base environment
+};
+```
+
+This is the production environment: `/src/environments/environment.prod.ts`:
+```typescript
+import { baseEnvironment } from './environment.base';
+
+export const environment: IEnvironment = {
+  ...baseEnvironment,
+  production: true,
+};
+```
+
+Make sure to update `angular.json` if you add/remove environment files.
+
+After all is defined, run/build the app for different environments by specifying the environment in `ng` command: `ng build --env=prod` for production build; `ng build --env=staging` for staging build; `ng build` will use default `dev` environment.
+
+## Formatting and Naming
+
+### In Templates
+
+For the purposes of future-proofing and avoiding conflicts with other libs, prefix all component selectors with something unique/app-specific. Specify the prefix in `angular.json` like so: `"prefix": "my-app",`.
+
+----
+
+Try to fit all inputs, outputs and regular HTML attributes in one line. If there is simply too many of them, break it up like so:
+
+```html
+<!-- bad -->
+<my-app-component *ngFor="let item of items" class="foo" [bar]="something ? 'foo' : 'oof'" (click)="onClick($event)">
+  <div>Transcluded Content</div>
+</my-app-component>
+
+<!-- bad -->
+<my-app-component
+  *ngFor="let item of items"
+  class="foo"
+  [bar]="something ? 'foo' : 'oof'"
+  (click)="onClick($event)">
+  <div>Transcluded Content</div>
+</my-app-component>
+
+<!-- maybe could be good? -->
+<my-app-component
+  *ngFor="let item of items"
+  class="foo"
+  [bar]="something ? 'foo' : 'oof'"
+  (click)="onClick($event)">
+
+  <div>Transcluded Content</div>
+</my-app-component>
+
+<!-- good -->
+<my-app-component
+  *ngFor="let item of items"
+  class="foo"
+  [bar]="something ? 'foo' : 'oof'"
+  (click)="onClick($event)"
+>
+  <div>Transcluded Content</div>
+</my-app-component>
+```
+
+----
+
+If you are passing data to a component, use property binding only if necessary:
+
+```html
+<!-- bad -->
+<my-app-component [name]="'Steve'"></my-app-component>
+
+<!-- good -->
+<my-app-component name="Steve"></my-app-component>
+```
+
+---
+
+Prefix output handlers with `on`:
```

Comment:
Maybe mention that handlers for very generic outputs and events (like `click`) could have a bit more descriptive names, like `onLoginButtonClick()` (or without the `Button` part)  instead of just `onClick()`.